### ISSUE TITLE
Encapsulation rules

### ISSUE THREAD

**Issue by MichaelClerx** [cellml/cellml-specification#40]
https://libcellml-tutorials.readthedocs.io/en/pr344_documentation/reference/specB14.html

> 1. Every component_ref element MUST contain a component attribute.
> The value of this attribute MUST be a valid CellML identifier, and MUST match the name attribute on a component element or an import component element in the CellML infoset.
> 
> 2. Every component_ref element MAY in turn contain one or more component_ref element children.
>
> 3. A component_ref element which is an immediate child of an encapsulation element MUST each contain at least one component_ref element child.

- Not sure why we need rule 3. Without it you could explicitly state "x has no children"
- Should there be a rule saying "a component_ref attribute must have a value that's unique within the encapsulation element" ? This can be handled syntactically, and prevents you from saying "x is a child of both y and z"

(Now you have to conclude that yourself from the later interpretation section " For the purposes of this specification, there SHALL be a “conceptual encapsulation digraph” in which there is EXACTLY one node for every component in the CellML model. Therefore the encapsulation digraph will not contain any loops.")

**Comment by MichaelClerx** [comment]
Also, not sure why we specify that an encapsulation element can't be empty. We allow empty models and components.

**Comment by MichaelClerx** [comment]
Earlier we have sometimes referred to the interpretation section, and added a rule saying the elements/attributes MUST meet the constraints set out there

**Comment by kerimoyle** [comment]
> Not sure why we need rule 3. Without it you could explicitly state "x has no children"

I don't understand what you mean?  The rule is meant to say "`encapsulation` must have children and grandchildren"?

> Earlier we have sometimes referred to the interpretation section, and added a rule saying the elements/attributes MUST meet the constraints set out there

We can do that :) I'll add a link to this section: https://cellml-specification-dev.readthedocs.io/en/latest/reference/specC9_interpretation_of_encapsulation.html 
 
> B14: A component_ref element information item (referred to in this specification as a component_ref element) is an element in the CellML namespace with a local name equal to component_ref, and which appears as a child of an encapsulation element.

See: https://cellml-specification-dev.readthedocs.io/en/latest/reference/specB14.html#the-component-ref-element

I wonder if this should also say it can be a child of another component_ref item too?

**Comment by MichaelClerx** [comment]
>>  Not sure why we need rule 3. Without it you could explicitly state "x has no children"
>
> I don't understand what you mean? The rule is meant to say "encapsulation must have children and grandchildren"?

Yeah and I wonder why we need that rule :-)
I'd think writing `<encapsulation />` would say "nothing in this model is encapsulated"
Or `<encapsulation><component_ref component="A" /></encapsulation>` would say "A has no children"

**Comment by MichaelClerx** [comment]
> I wonder if this should also say it can be a child of another component_ref item too?

I think so! Or else change to "appears within an encapsulation element" ?

**Comment by kerimoyle** [comment]
> Or <encapsulation><component_ref component="A" /></encapsulation> would say "A has no children" 
.. also means nothing in this model is encapsulated ...

**Comment by MichaelClerx** [comment]
:D
OK so something like `<encapsulation><component_ref component="A" />...</encapsulation>` where `...` denotes a nice hierarchy that doesn't involve A

**Comment by kerimoyle** [comment]
I quite like this idea, and I think it's more consistent with other parts of the spec (empty models, components, maths blocks etc).  That whatever you specify in (a single) encapsulation is what you get.  So empty ones are fine, ones with a single generation are fine (both effectively mean there is no real hierarchy), ones as in the example above are fine ... it's a lot clearer this way, and easier to write about ...

In terms of the specification this means it could change to:

> **B 13.1.1** Every encapsulation element **MAY** (cf MUST) contain one or more component_ref elements.

> B 14. The component_ref element
A component_ref element information item (referred to in this specification as a component_ref element) is an element in the CellML namespace with a local name equal to component_ref, and which appears as a child of an encapsulation element **or a component_ref element**.

> 14.1. Specific information items
14.1.1 Every component_ref element MUST contain a component attribute.
14.1.2 The value of this attribute MUST be a valid CellML identifier, and MUST match the name attribute on a component element or an import component element in the CellML infoset.
14.1.3 Every component_ref element MAY in turn contain one or more component_ref element children.
**Delete(A component_ref element which is an immediate child of an encapsulation element MUST each contain at least one component_ref element child)**

Basically this means that the two last examples under the "see more" link [here](https://cellml-specification-dev.readthedocs.io/en/latest/reference/specB13.html) become valid CellML.

**Comment by kerimoyle** [comment]
Any changes here? @nickerso @agarny @hsorby ?

**Comment by kerimoyle** [comment]
Ping @MichaelClerx @nickerso @agarny @hsorby

**Comment by hsorby** [comment]
I agree, I can't see a reason to stop people writing empty encapsulations.  I don't think we should invalidate the model because of it.  As far as I can tell there are no ramifications for declaring an empty encapsulation.

(Couldn't follow that link)

**Comment by hsorby** [comment]
I think `B 13.1.1` should change to encapsulation elements MUST only have children that are component_ref. (or whatever the proper way of saying that is.)

**Comment by kerimoyle** [comment]
Try this one instead:  https://cellml-specification.readthedocs.io/en/cellml-2-drafting/reference/formal_and_informative/specB14.html

**Comment by hsorby** [comment]
I don't see any semantic difference between saying that component A has no children and not saying component A has no children.

**Comment by kerimoyle** [comment]
> I don't see any semantic difference between saying that component A has no children and not saying component A has no children.

There's not ... Maybe we add a statement saying that the default is to be without children?  ie: this does nothing and could be left out:
```
<encapsulation>
  <component_ref component="A" />
</encapsulation>
```
It's just that at present this is invalid, so that would be a change.

**Comment by hsorby** [comment]
I don't think the specification needs such a statement.  Such a statement shouldn't be part of the normative spec.

**Comment by kerimoyle** [comment]
OK, fair enough.  I'll put something in the informative one instead.

**Comment by kerimoyle** [comment]
Closed by #96
